src/libraries/System.Private.CoreLib/src/System/IO/Strategies/FileStreamValueTaskSource.cs

```diff
@@ -0,0 +1,252 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks.Sources;
+
+namespace System.IO.Strategies
+{
+    internal sealed partial class AsyncWindowsFileStreamStrategy : WindowsFileStreamStrategy
+    {
+        /// <summary>
+        /// Type that helps reduce allocations for FileStream.ReadAsync and FileStream.WriteAsync.
+        /// </summary>
+        private unsafe class FileStreamValueTaskSource : IValueTaskSource<int>, IValueTaskSource
+        {
+            internal static readonly IOCompletionCallback s_ioCallback = IOCallback;
+
+            private readonly AsyncWindowsFileStreamStrategy _strategy;
+            private readonly int _numBufferedBytes;
+
+            private ManualResetValueTaskSourceCore<int> _source; // mutable struct; do not make this readonly
+            private NativeOverlapped* _overlapped;
+            private CancellationTokenRegistration _cancellationRegistration;
+            private long _result; // Using long since this needs to be used in Interlocked APIs
+#if DEBUG
+            private bool _cancellationHasBeenRegistered;
+#endif
+
+            public static FileStreamValueTaskSource Create(
+                AsyncWindowsFileStreamStrategy strategy,
+                PreAllocatedOverlapped? preallocatedOverlapped,
+                int numBufferedBytes,
+                ReadOnlyMemory<byte> memory)
+            {
+                // If the memory passed in is the strategy's internal buffer, we can use the base AwaitableProvider,
+                // which has a PreAllocatedOverlapped with the memory already pinned.  Otherwise, we use the derived
+                // MemoryAwaitableProvider, which Retains the memory, which will result in less pinning in the case
+                // where the underlying memory is backed by pre-pinned buffers.
+                return preallocatedOverlapped != null &&
+                       MemoryMarshal.TryGetArray(memory, out ArraySegment<byte> buffer) &&
+                       preallocatedOverlapped.IsUserObject(buffer.Array) ?
+                            new FileStreamValueTaskSource(strategy, preallocatedOverlapped, numBufferedBytes, buffer.Array) :
+                            new MemoryAwaitableProvider(strategy, numBufferedBytes, memory);
+            }
+
+            protected FileStreamValueTaskSource(
+                AsyncWindowsFileStreamStrategy strategy,
+                PreAllocatedOverlapped? preallocatedOverlapped,
+                int numBufferedBytes,
+                byte[]? bytes)
+            {
+                _strategy = strategy;
+                _numBufferedBytes = numBufferedBytes;
```

Comment:
Consider removing this field since is always set to zero. FWIW there was one case in legacy code that was setting this field to a value other than zero; since FileStreamValueTaskSource is only used in non-legacy, we should be fine with just removing it.

Reply:
If it's not meaningful, we definitely shouldn't have it :)